File: masar_miraaya_shipping/custom/pick_list/pick_list.py
import frappe
from frappe import _
from frappe.utils import now_datetime, time_diff_in_seconds, format_duration


def validate(self, method):
    if not self.custom_picking_datetime and self.docstatus == 0:
        self.custom_picking_datetime = now_datetime()
def on_submit(self, method):
    validate_picked_qty(self)       
    stop_picking_timer(self)

def validate_picked_qty(self):
    for item in self.locations:
        if item.picked_qty <= 0:
            frappe.throw(_(f"Picked Quantity must be greater than zero for item {item.item_code} in row {item.idx}"))
    if not self.custom_assigned:
        frappe.throw(_("Please assign the picker before submitting the Pick List."))


@frappe.whitelist()
def start_picking_timer(pick_list_name):
    doc = frappe.get_doc("Pick List", pick_list_name)
    
    if not doc.custom_picking_datetime:
        doc.custom_picking_datetime = now_datetime()
        if not doc.custom_picking_user:
            doc.custom_picking_user = frappe.session.user
        doc.save(ignore_permissions=True)
        
    return {
        "success": True,
        "start_time": doc.custom_picking_datetime
    }
    
@frappe.whitelist()
def start_packing_timer(pick_list_name):
    doc = frappe.get_doc("Pick List", pick_list_name)
    
    if not doc.custom_packing_datetime:
        doc.custom_packing_datetime = now_datetime()
        if not doc.custom_packing_user:
            doc.custom_packing_user = frappe.session.user
        doc.save(ignore_permissions=True)
        
    return {
        "success": True,
        "start_time": doc.custom_packing_datetime
    }
    
def stop_picking_timer(self):
    if self.custom_picking_datetime and not self.custom_picking_end_datetime:
        self.custom_picking_end_datetime = now_datetime()
        self.custom_picking_duration = time_diff_in_seconds(self.custom_picking_end_datetime, self.custom_picking_datetime)
        if not self.custom_picking_user:
            self.custom_picking_user = frappe.session.user
        self.save(ignore_permissions=True)

File: masar_miraaya_shipping/custom/pick_list/pick_list.js
frappe.ui.form.on("Pick List", {
    onload: function(frm) {
        if (frm.doc.__islocal != 1 && frm.doc.docstatus === 0 && !frm.doc.custom_picking_end_datetime) {
            start_picking_timer(frm);
        }
        if (frm.doc.docstatus === 1 && !frm.doc.custom_packing_end_datetime) {
            start_packing_timer(frm);
        }
    }
});

function start_picking_timer(frm) {
    frappe.call({
        method: "masar_miraaya_shipping.custom.pick_list.pick_list.start_picking_timer",
        args: {
            pick_list_name: frm.doc.name
        },
        callback: function(r) {
            if (r.message && r.message.success) {
                frm.refresh_field("custom_picking_datetime");
                frm.refresh_field("custom_picking_user");
                frm.refresh_field("custom_picking_end_datetime");
            }
        }
    });
}

function start_packing_timer(frm) {
    frappe.call({
        method: "masar_miraaya_shipping.custom.pick_list.pick_list.start_packing_timer",
        args: {
            pick_list_name: frm.doc.name
        },
        callback: function(r) {
            if (r.message && r.message.success) {
                frm.refresh_field("custom_packing_datetime");
                frm.refresh_field("custom_packing_user");
                frm.refresh_field("custom_packing_end_datetime");
            }
        }
    });
}

frappe.ui.form.on('Pick List', {

    refresh(frm) {
        if (frm.fields_dict.custom_scan_packaging_barcode) {
            frm.fields_dict.custom_scan_packaging_barcode.$input.on("keypress", function(e) {
                if (e.which === 13) {
                    frm.trigger("custom_scan_packaging_barcode");
                }
            });
        }
    },

    custom_scan_packaging_barcode(frm) {

        if (!frm.doc.custom_scan_packaging_barcode) return;

        let barcode = frm.doc.custom_scan_packaging_barcode;

        frappe.db.get_value("Item", { name: barcode }, ["name", "item_name"])
            .then(r => {

                if (r.message) {
                    process_item(frm, r.message.name, r.message.item_name);
                    return;
                }

                frappe.db.get_value("Item Barcode", { barcode: barcode }, "parent")
                    .then(res => {

                        if (!res.message) {
                            frappe.msgprint("Barcode not found");
                            frm.set_value("custom_scan_packaging_barcode", "");
                            return;
                        }

                        let item_code = res.message.parent;

                        frappe.db.get_value("Item", item_code, ["name", "item_name"])
                            .then(item => {
                                process_item(frm, item.message.name, item.message.item_name);
                            });
                    });
            });
    }
});


function process_item(frm, item_code, item_name) {

    let existing = frm.doc.custom_packaging_items.find(d => d.item_code === item_code);

    if (existing) {
        frappe.show_alert({
            message: `${item_name} already added`,
            indicator: "orange"
        });
    } else {

        let child = frm.add_child("custom_packaging_items");

        child.item_code = item_code;
        child.item_name = item_name;

        frm.refresh_field("custom_packaging_items");

        frappe.show_alert({
            message: `${item_name} added`,
            indicator: "green"
        });
    }

    frm.set_value("custom_scan_packaging_barcode", "");
}